README.md
# Helix Word to Markdown

> A service that converts Word documents to markdown.

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-word2md.svg)](https://codecov.io/gh/adobe/helix-word2md)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-word2md.svg)](https://circleci.com/gh/adobe/helix-word2md)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-word2md.svg)](https://github.com/adobe/helix-word2md/blob/master/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-word2md.svg)](https://github.com/adobe/helix-word2md/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-word2md.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-word2md)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release) [![Greenkeeper badge](https://badges.greenkeeper.io/adobe/helix-word2md.svg)](https://greenkeeper.io/)

## Installation

## Usage

```bash
curl https://adobeioruntime.net/api/v1/web/helix/helix-services/word2md@v1?shareLink=...&path=...
```

## Testing

You can browse the OneDrive integration using _browser.js_:

1. start with: `npm run start`.
2. open web browser at `http://localhost:3000/`.
3. sign in with microsoft.
3. copy-paste share-link of a shared folder (see below) and click the `list` button.


## Authentication

The action authenticates against OneDrive using an oauth2 refresh token. In case you need to
create a new one, use the _browser.js_ to generate a `tokens.json`.

## Development

### Deploying Helix Word to Markdown

Deploying a Helix service requires the `wsk` command-line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to `master` that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/gdocs2md@ci<num>` and tagged with the CI build number.

## Appendix

### Setting up a shared folder in OneDrive

1. Create a new folder in OneDrive. eg `helix`.
2. Click on the (i) button and then _Manage Access_.
![](./docs/step1.png)
3. Add the email address of the user you want to share. For the integration with helix-pages use: `helix.integration@gmail.com`.
![](./docs/step2.png)
4. Click on _Copy Link_ and then _Copy_ in order to copy the link to your clipboard.
![](./docs/step3.png)
5. Create a word document in your folder. eg: `test.docx`.
6. open a browser at `https://adobeioruntime.net/api/v1/web/helix/helix-services/word2md@v1?shareLink=<your-share-link>&path=/test.docx`.



CHANGELOG.md
## [1.0.1](https://github.com/adobe/helix-word2md/compare/v1.0.0...v1.0.1) (2019-10-02)


### Bug Fixes

* **deps:** pin dependency mammoth to 1.4.8 ([#10](https://github.com/adobe/helix-word2md/issues/10)) ([5ed7473](https://github.com/adobe/helix-word2md/commit/5ed7473))
* **monitoring:** use new automation tooling ([e339b62](https://github.com/adobe/helix-word2md/commit/e339b62))

# 1.0.0 (2019-10-01)


### Bug Fixes

* **deps:** pin dependencies ([f6f976c](https://github.com/adobe/helix-word2md/commit/f6f976c))


### Features

* **docx:** add docx to md converter ([cd7549b](https://github.com/adobe/helix-word2md/commit/cd7549b))

## [1.2.1](https://github.com/adobe/helix-googledocs2md/compare/v1.2.0...v1.2.1) (2019-09-20)


### Bug Fixes

* **table:** ensure space before table ([#16](https://github.com/adobe/helix-googledocs2md/issues/16)) ([26ed2d1](https://github.com/adobe/helix-googledocs2md/commit/26ed2d1))

# [1.2.0](https://github.com/adobe/helix-googledocs2md/compare/v1.1.1...v1.2.0) (2019-09-18)


### Features

* **docs2md:** adding table support ([#13](https://github.com/adobe/helix-googledocs2md/issues/13)) ([5ab703f](https://github.com/adobe/helix-googledocs2md/commit/5ab703f))

## [1.1.1](https://github.com/adobe/helix-googledocs2md/compare/v1.1.0...v1.1.1) (2019-09-18)


### Bug Fixes

* **docs2md:** fix formatting issues with headings ([#12](https://github.com/adobe/helix-googledocs2md/issues/12)) ([ecafe14](https://github.com/adobe/helix-googledocs2md/commit/ecafe14))

# [1.1.0](https://github.com/adobe/helix-googledocs2md/compare/v1.0.0...v1.1.0) (2019-09-17)


### Features

* **google:** add google docs to markdown coverter ([1079ea1](https://github.com/adobe/helix-googledocs2md/commit/1079ea1))

# 1.0.0 (2019-09-10)


### Features

* **google:** fetch markdown via google web app ([1de8bc9](https://github.com/adobe/helix-googledocs2md/commit/1de8bc9))

## [1.2.1](https://github.com/adobe/helix-service/compare/v1.2.0...v1.2.1) (2019-09-03)


### Bug Fixes

* **package:** update @adobe/helix-status to version 4.2.1 ([f7ab47a](https://github.com/adobe/helix-service/commit/f7ab47a))

# [1.2.0](https://github.com/adobe/helix-service/compare/v1.1.2...v1.2.0) (2019-08-22)


### Features

* **monitoring:** add helix-status library to enable monitoring ([ef72300](https://github.com/adobe/helix-service/commit/ef72300))

## [1.1.2](https://github.com/adobe/helix-service/compare/v1.1.1...v1.1.2) (2019-07-24)


### Bug Fixes

* **package:** fix npm test script in order to provide information about failing tests ([80e7fe2](https://github.com/adobe/helix-service/commit/80e7fe2)), closes [#13](https://github.com/adobe/helix-service/issues/13)

## [1.1.1](https://github.com/adobe/helix-service/compare/v1.1.0...v1.1.1) (2019-06-17)


### Bug Fixes

* **package.json:** remove redundant istanbul dependency and use latest nyc ([b29f97d](https://github.com/adobe/helix-service/commit/b29f97d)), closes [#4](https://github.com/adobe/helix-service/issues/4)

# [1.1.0](https://github.com/adobe/helix-service/compare/v1.0.1...v1.1.0) (2019-06-12)


### Features

* **action:** turn action into a web action ([f41f212](https://github.com/adobe/helix-service/commit/f41f212))

## [1.0.1](https://github.com/adobe/helix-service/compare/v1.0.0...v1.0.1) (2019-06-12)


### Bug Fixes

* **build:** add missing dependency ([aa163d7](https://github.com/adobe/helix-service/commit/aa163d7))

# 1.0.0 (2019-06-12)


### Bug Fixes

* **build:** add wsk property for release tracking ([9e36a10](https://github.com/adobe/helix-service/commit/9e36a10))
* **build:** increase version number to get a release ([f04ab95](https://github.com/adobe/helix-service/commit/f04ab95))


test/browser.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
/* eslint-disable no-console */
const crypto = require('crypto');
const fs = require('fs');
const express = require('express');
const cookieParser = require('cookie-parser');
const OneDrive = require('./../src/OneDrive.js');

require('dotenv').config();

const redirectUri = 'http://localhost:3000/token';

function createOneDriveClient() {
  const {
    AZURE_WORD2MD_CLIENT_ID: clientId,
    AZURE_WORD2MD_CLIENT_SECRET: clientSecret,
    AZURE_WORD2MD_REFRESH_TOKEN: refreshToken,
  } = process.env;

  let tokens = {};
  try {
    tokens = JSON.parse(fs.readFileSync('tokens.json', 'utf-8'));
  } catch (e) {
    // ignore
  }

  const {
    accessToken,
    expiresOn,
  } = tokens;

  return new OneDrive({
    clientId,
    clientSecret,
    refreshToken,
    accessToken,
    expiresOn,
  });
}
const drive = createOneDriveClient();

// register event handle to write back tokens.
drive.on('tokens', (tokens) => {
  fs.writeFileSync('tokens.json', JSON.stringify(tokens, null, 2), 'utf-8');
  console.log('updated "tokens.json" file.');
});

// setup express app

async function auth(req, res) {
  crypto.randomBytes(48, (ex, buf) => {
    const state = buf.toString('base64').replace(/\//g, '_').replace(/\+/g, '-');
    const authorizationUrl = drive.createLoginUrl(redirectUri, state);
    res.cookie('authstate', state);
    res.redirect(authorizationUrl);
  });
}

async function token(req, res) {
  if (req.cookies.authstate !== req.query.state) {
    res.send('error: state does not match');
    return;
  }
  try {
    await drive.aquireToken(redirectUri, req.query.code);
    res.redirect('/');
  } catch (e) {
    console.error(e);
    res.send(`error: ${e.message}`);
  }
}

async function root(req, res) {
  if (!drive.authenticated) {
    res.send('unauthorized. <a href="/auth">sign in</a>');
    return;
  }
  try {
    const result = await drive.me();
    res.setHeader('content-type', 'text/html; charset=utf-8');
    const html = [
      `welcome <b>${result.displayName}</b><br>`,
      '<form method="get" action="/list">',
      'Share link: <input name="l" size=40 value="https://adobe-my.sharepoint.com/personal/tripod_adobe_com/Documents/helix-content?csf=1&e=Fz6r5Z"><br>',
      '<button>list</button><br>',
      '</form>',
    ].join('\n');
    res.end(html);
  } catch (e) {
    console.error(e);
    res.status(500).send('Something broke!');
  }
}

async function listDocuments(req, res) {
  const { l } = req.query;
  if (!l) {
    res.end('no share lin provided.');
    return;
  }
  try {
    const pa = req.path.replace(/\/+$/, '');
    const rootItem = await drive.getDriveItemFromShareLink(l);
    const result = await drive.listChildren(rootItem, req.path);
    const list = result.value.map((entry) => {
      const e = {
        name: entry.name,
        id: entry.id,
        link: `/${entry.folder ? 'list' : 'md'}${pa}/${entry.name}`,
      };
      return e;
    });

    const html = list.map((e) => `<a href="${e.link}?l=${encodeURIComponent(l)})">${e.name}</a><br>`).join('\n');
    res.setHeader('content-type', 'text/html; charset=utf-8');
    res.end(html);
  } catch (e) {
    console.error(e);
    res.status(500)
      .send('Something broke!');
  }
}

async function md(req, res) {
  const { l } = req.query;
  if (!l) {
    res.end('no share lin provided.');
    return;
  }
  try {
    const rootItem = await drive.getDriveItemFromShareLink(l);
    const result = await drive.getDriveItem(rootItem, req.path, true);
    res.setHeader('content-type', 'application/octet-stream');
    res.send(result);
  } catch (e) {
    console.error(e);
    res.status(500)
      .send('Something broke!');
  }
}

function asyncHandler(fn) {
  return (req, res, next) => (Promise.resolve(fn(req, res, next)).catch(next));
}

const app = express();
app.use(cookieParser());


// generate login redirect
app.get('/auth', asyncHandler(auth));

// After consent is granted AAD redirects here.
app.get('/token', asyncHandler(token));


app.use('/list', express.Router().get('*', asyncHandler(listDocuments)));
app.use('/md', express.Router().get('*', asyncHandler(md)));
app.get('/', asyncHandler(root));

app.listen(3000, () => console.log(`Example app listening on port ${3000}!`));


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

const path = require('path');
const fse = require('fs-extra');
const assert = require('assert');
const nock = require('nock');
const { main } = require('../src/index.js');

const DUMMY_PARAMS = {
  AZURE_WORD2MD_CLIENT_ID: 'dummy',
  AZURE_WORD2MD_CLIENT_SECRET: 'dummy',
  AZURE_WORD2MD_REFRESH_TOKEN: 'dummy',
  shareLink: 'abcd',
  path: '/foo',
};

describe('Action Tests', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  ['AZURE_WORD2MD_CLIENT_ID', 'AZURE_WORD2MD_CLIENT_SECRET', 'AZURE_WORD2MD_REFRESH_TOKEN', 'shareLink', 'path'].forEach((name, idx, names) => {
    it(`action fails if ${name} parameter is missing`, async () => {
      const params = {};
      for (let i = 0; i < idx; i += 1) {
        params[names[i]] = 'dummy';
      }
      const result = await main(params);
      assert.deepEqual(result, {
        body: `${name} parameter missing.`,
        headers: {
          'Cache-Control': 'no-store, private, must-revalidate',
        },
        statusCode: 500,
      });
    });
  });

  it('action returns content from google docs', async () => {
    const docx = await fse.readFile(path.resolve(__dirname, 'fixtures', 'document.docx'));
    const docsMd = await fse.readFile(path.resolve(__dirname, 'fixtures', 'doc.md'), 'utf-8');
    nock('https://login.windows.net')
      .post('/common/oauth2/token?api-version=1.0')
      .reply(200, {
        token_type: 'Bearer',
        refresh_token: 'dummy',
        access_token: 'dummy',
        expires_in: 81000,
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/shares/u!YWJjZA=/driveItem')
      .reply(200, {
        id: 'docid',
        parentReference: {
          driveId: 'driveid',
        },
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/drives/driveid/items/docid:/foo:/content')
      .reply(200, docx);

    const result = await main(DUMMY_PARAMS);
    assert.deepEqual(result, {
      statusCode: 200,
      body: docsMd,
      headers: {
        'Cache-Control': 'no-store, private, must-revalidate',
        'Content-Length': docsMd.length,
        'Content-Type': 'text/plain',
      },
    });
  });

  it('action returns content from google docs with folder', async () => {
    const docx = await fse.readFile(path.resolve(__dirname, 'fixtures', 'document.docx'));
    const docsMd = await fse.readFile(path.resolve(__dirname, 'fixtures', 'doc.md'), 'utf-8');
    nock('https://login.windows.net')
      .post('/common/oauth2/token?api-version=1.0')
      .reply(200, {
        token_type: 'Bearer',
        refresh_token: 'dummy',
        access_token: 'dummy',
        expires_in: 81000,
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/shares/u!YWJjZA=/driveItem')
      .reply(200, {
        id: 'docid',
        parentReference: {
          driveId: 'driveid',
        },
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/drives/driveid/items/docid:/foo/bar:/content')
      .reply(200, docx);


    const result = await main({
      ...DUMMY_PARAMS,
      path: '/foo/bar',
    });
    assert.deepEqual(result, {
      statusCode: 200,
      body: docsMd,
      headers: {
        'Cache-Control': 'no-store, private, must-revalidate',
        'Content-Length': docsMd.length,
        'Content-Type': 'text/plain',
      },
    });
  });

  it('action handles 404 if file not found', async () => {
    nock('https://login.windows.net')
      .post('/common/oauth2/token?api-version=1.0')
      .reply(200, {
        token_type: 'Bearer',
        refresh_token: 'dummy',
        access_token: 'dummy',
        expires_in: 81000,
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/shares/u!YWJjZA=/driveItem')
      .reply(200, {
        id: 'docid',
        parentReference: {
          driveId: 'driveid',
        },
      });
    nock('https://graph.microsoft.com/v1.0')
      .get('/drives/driveid/items/docid:/not_exists:/content')
      .reply(404);

    const result = await main({
      ...DUMMY_PARAMS,
      path: '/not_exists',
    });
    assert.deepEqual(result, {
      body: 'error while converting document',
      headers: {
        'Cache-Control': 'no-store, private, must-revalidate',
      },
      statusCode: 404,
    });
  });
});


test/docs2md.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

const fse = require('fs-extra');
const path = require('path');
const assert = require('assert');
const docs2md = require('../src/docs2md');

describe('Docs to markdown converter', () => {
  it('converts a google docs json correctly', async () => {
    const doc = await fse.readJson(path.resolve(__dirname, 'fixtures', 'doc.json'));
    const expected = await fse.readFile(path.resolve(__dirname, 'fixtures', 'doc.md'), 'utf-8');
    const actual = docs2md(doc);
    assert.equal(actual, expected);
  });
});


src/docs2md.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-disable no-control-regex */

/**
 * Google Document
 * @external Document
 * @see {@link https://developers.google.com/docs/api/reference/rest/v1/document}
 */

/**
 * Google Document - Structural Element
 * @external StructuralElement
 * https://developers.google.com/docs/api/reference/rest/v1/documents#StructuralElement
 */

/**
 * Text style to heading mapping.
 */
const HEADINGS = {
  TITLE: '# ',
  SUBTITLE: '### ',
  HEADING_1: '# ',
  HEADING_2: '## ',
  HEADING_3: '### ',
  HEADING_4: '#### ',
  HEADING_5: '##### ',
  HEADING_6: '###### ',
};

/**
 * Text style to markdown code mapping
 */
const TEXT_STYLES = {
  bold: {
    name: 'bold',
    on: '**',
    off: '**',
  },
  italic: {
    name: 'italic',
    on: '_',
    off: '_',
  },
  underline: {
    name: 'underline',
    on: '<u>',
    off: '</u>',
  },
  strikethrough: {
    name: 'strikethrough',
    on: '<del>',
    off: '</del>',
  },
  SUPERSCRIPT: {
    name: 'baselineOffset',
    on: '<sup>',
    off: '</sup>',
  },
  SUBSCRIPT: {
    name: 'baselineOffset',
    on: '<sub>',
    off: '</sub>',
  },
};

/**
 * Monospace fonts used in google docs
 */
const CODE_FONTS = {
  'Courier New': true,
  'Source Code Pro': true,
  VT323: true,
  Consolas: true,
  Courier: true,
  'Nanum Gothic Coding': true,
  Cousine: true,
};

/**
 * Checks if the font specified by the text style is a font used to format code.
 * @param {Schema$TextStyle} textStyle
 * @returns {boolean} {@code true} if the font is a code font.
 */
function isCodeFont({ weightedFontFamily }) {
  if (!weightedFontFamily) {
    return false;
  }
  const { fontFamily } = weightedFontFamily;
  return (fontFamily in CODE_FONTS || fontFamily.match(/\sMono/));
}

/**
 * Processes an array of structural elements.
 *
 * @param {Document} doc - the google document
 * @param {StructuralElement[]} structuralElements - the content of the document body or table cell
 * @param {boolean} isTableCell - indicates whether a table cell is processed.
 * @returns the markdown representation.
 */
function processStructuralElements(doc, structuralElements, isTableCell = false) {
  let md = '';
  let wasBullet = false;
  let wasHeading = true;
  const blockCode = [];
  const lf = isTableCell ? ' ' : '\n'; // ignore line feeds in table cells.

  const escapeCode = (txt) => (txt.indexOf('`') >= 0 ? `\`\`${txt}\`\`` : `\`${txt}\``);

  // code is gathered outside the paragraph loop, so that multiple code paragraphs generate one
  // multiline code block. this is what happens when copy-pasting code into a google docs.
  const flushCode = (code) => {
    if (code.length === 0) {
      return '';
    }
    let txt = '';
    if (code.length === 1) {
      txt = escapeCode(code[0]);
    } else if (isTableCell) {
      txt = escapeCode(code.join(' '));
    } else {
      txt = `\`\`\`\n${code.join('\n')}\n\`\`\``;
    }
    code.splice(0, code.length);
    return txt;
  };

  structuralElements.forEach((el, elIdx) => {
    if ('paragraph' in el) {
      const par = el.paragraph;
      const style = par.paragraphStyle.namedStyleType;
      const heading = HEADINGS[style] || '';
      let currentStyles = [];
      let trailing = ''; // remembers trailing white space
      const isBullet = 'bullet' in par;
      let numTexts = 0;

      // flush code in bullet
      if (isBullet || wasBullet) {
        const cd = flushCode(blockCode);
        if (cd) {
          md += `${cd}${lf}`;
        }
      }

      if (isBullet) {
        if (!wasBullet) {
          md += lf;
        }
        // assume no bullet with heading
        const blt = par.bullet;
        const lvl = blt.nestingLevel || 0;
        md += `${'    '.repeat(lvl)}`;
        const list = doc.lists[blt.listId];
        let symbol = '- ';
        if (list) {
          const { nestingLevels } = list.listProperties;
          const level = nestingLevels[blt.nestingLevel || 0];
          if (level) {
            if (level.glyphType) {
              symbol = '1. ';
            }
          }
        }
        md += symbol;
      } else {
        // insert a new line before new heading, between paragraphs and after lists
        // eslint-disable-next-line no-lonely-if
        if ((heading && md) || (!wasHeading && blockCode.length === 0) || wasBullet) {
          md += lf;
        }
      }
      wasHeading = heading;
      wasBullet = isBullet;

      let p = '';
      let pCode = [];
      par.elements.forEach((pel) => {
        if ('textRun' in pel) {
          const tr = pel.textRun;
          const ts = tr.textStyle;
          const isCode = isCodeFont(ts);

          // remove trailing new line
          const content = tr.content.replace(/\n$/gm, '');
          if (content) {
            numTexts += 1;
          }

          // split along the soft breaks
          content.split(/\u000b/gu).forEach((text, idx) => {
            if (isCode) {
              if (pCode.length > 0) {
                if (idx === 0) {
                  pCode[pCode.length - 1] += text; // multiple code formats on the same
                } else {
                  pCode.push(text);
                }
              } else {
                // turn styles off
                p += currentStyles.reduceRight((c, s) => c + s.off, '');
                currentStyles = [];
                p += trailing;
                trailing = ''; // all whitespace in code are relevant
                pCode.push(text);
              }
            } else {
              if (blockCode.length > 0) {
                blockCode.push(...pCode);
                pCode.splice(0, pCode.length);
                // flush any block of code before this paragraph
                const cd = flushCode(blockCode);
                md += cd;
                if (heading && cd) {
                  md += `${lf}${lf}`;
                }
              }

              // special case, if code was just before a softbreak
              if (pCode.length === 2 && !pCode[1]) {
                p += `${escapeCode(pCode[0])}  ${lf}`;
                pCode = [];
              }
              p += flushCode(pCode);

              // markdown does not allow whitespace between format characters and words
              // so extract the leading and trailing whitespace
              // eslint-disable-next-line prefer-const
              let [, pre, mid, post] = /^(\s*)(.*?)(\s*?)$/.exec(text);

              // first, turn styles off that are not in text style
              const activeStyles = {};
              for (let i = currentStyles.length - 1; i >= 0; i -= 1) {
                const s = currentStyles[i];
                if (!ts[s.name]) {
                  p += s.off;
                  currentStyles.splice(i, 1);
                } else {
                  activeStyles[s.name] = true;
                }
              }
              // then add the previous trailing whitespace
              p += trailing;

              // if there was a manual break before, add the new line
              if (idx > 0) {
                p += `  ${lf}`;
              }

              // leading whitespace
              p += pre;

              // don't turn on styles for empty text
              if (mid.length === 0) {
                trailing = post;
                return;
              }

              // turn styles on
              Object.keys(ts).forEach((s) => {
                if (ts[s]) {
                  if (s === 'underline' && ts.link) {
                    // ignore underline for links
                    mid = `[${mid}](${ts.link.url})`;
                    return;
                  }
                  if (activeStyles[s]) {
                    return;
                  }
                  if (s === 'baselineOffset') {
                    // for sub- and superscript, use the baselineOffset value as style name.
                    // eslint-disable-next-line no-param-reassign
                    s = ts[s];
                  }
                  const mdStyle = TEXT_STYLES[s];
                  if (mdStyle) {
                    p += mdStyle.on;
                    currentStyles.push(mdStyle);
                  }
                }
              });

              // add the mid text
              p += mid;

              // remember trailing ws
              trailing = post;
            }
          });
        } else if ('autoText' in pel) {
          // todo
        } else if ('pageBreak' in pel) {
          p += '\n\n---\n\n';
        } else if ('columnBreak' in pel) {
          // todo
        } else if ('footnoteReference' in pel) {
          // todo
        } else if ('horizontalRule' in pel) {
          p += '\n\n---\n\n';
        } else if ('equation' in pel) {
          // todo
        } else if ('inlineObjectElement' in pel) {
          const id = pel.inlineObjectElement.inlineObjectId;
          const inlineObject = doc.inlineObjects[id];
          const { embeddedObject } = inlineObject.inlineObjectProperties;
          if ('imageProperties' in embeddedObject) {
            numTexts += 1;
            const { imageProperties, title } = embeddedObject;
            p += trailing;
            trailing = ' ';
            p += `![${title || ''}](${imageProperties.contentUri})`;
          } else {
            // console.warn('unsupported embedded object ');
          }
        } else {
          // n/a - should not happen
        }
      });

      // end of paragraph
      if (pCode.length > 0) {
        if (numTexts > 1) {
          // inline code at end of paragraph.
          p += `${trailing}${flushCode(pCode)}`;
        } else {
          // single line code
          blockCode.push(...pCode);
        }
      }
      if (blockCode.length === 0) {
        // turn styles off
        p += currentStyles.reduceRight((c, s) => c + s.off, '');
        p += `${trailing}`;

        // ignore empty heading
        if (heading && numTexts > 0) {
          md += `${heading + p.replace('\n', '<br>')}${lf}`;
        } else {
          md += `${p}${lf}`;
        }
      }
    } else if ('sectionBreak' in el) {
      if (elIdx > 0) {
        // ignore the very first section break
        md += '\n---\n\n';
      }
    } else if ('table' in el) {
      // make sure there is an empty line before the table
      if (!md.endsWith('\n\n')) {
        md += '\n';
      }
      el.table.tableRows.forEach((row, rowIdx) => {
        let separator = '';
        row.tableCells.forEach((cell, cellIdx) => {
          if (cellIdx === 0) {
            md += '|';
          }
          const txt = processStructuralElements(doc, cell.content, true);
          md += `${txt}|`;

          // this is a bit a hack to get the alignments from the first row of cells
          if (rowIdx === 0) {
            separator += '|';
            const ps = cell.content[0]
              && cell.content[0].paragraph
              && cell.content[0].paragraph.paragraphStyle
              && cell.content[0].paragraph.paragraphStyle.alignment;
            if (ps === 'CENTER') {
              separator += ':';
            }
            separator += '-'.repeat(Math.min(txt.length, 10));
            if (ps === 'CENTER' || ps === 'END') {
              separator += ':';
            }
          }
        });
        if (separator) {
          md += `\n${separator}|`;
        }
        md += '\n';
      });
    } else {
      // n/a - should not happen
    }
  });
  md += flushCode(blockCode);
  return md;
}

/**
 * Converts a google docs json to markdown.
 * @see https://developers.google.com/docs/api/reference/rest/v1/documents#Document
 *
 * @param {Document} doc - the google document
 * @returns the markdown representation.
 */
function docs2markdown(doc) {
  return processStructuralElements(doc, doc.body.content);
}

module.exports = docs2markdown;


src/docx2md.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const mammoth = require('mammoth');

/**
 * Converts a docx to markdown.
 *
 * @param {Buffer} doc - the Word document
 * @returns the markdown representation.
 */
async function docx2markdown(doc) {
  const res = await mammoth.convertToMarkdown(doc, {
    styleMap: [
      "p[style-name='Code Block'] => code",
      "p[style-name='Inline Code'] => code",
    ],
  });
  if (res.messages.length > 0) {
    // eslint-disable-next-line no-console
    console.log(res.messages);
  }
  return res.value;
}

module.exports = docx2markdown;


src/index.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
/* eslint-disable no-console,no-param-reassign */
const { logger } = require('@adobe/openwhisk-action-utils');
const statusWrap = require('@adobe/helix-status').wrap;
const OneDrive = require('./OneDrive.js');
const docx2md = require('./docx2md.js');

/**
 * Generates an error response
 * @param {string} message - error message
 * @param {number} statusCode - error code.
 * @returns response
 */
function error(message, statusCode = 500) {
  return {
    statusCode,
    headers: {
      'Cache-Control': 'no-store, private, must-revalidate',
    },
    body: message,
  };
}

/**
 * Fetches the Google document via the API and converts it via docs2md.
 * @param params
 * @returns {Promise<response>}
 */
async function fetchViaAPI(params) {
  const { __ow_logger: log } = params;
  const {
    AZURE_WORD2MD_CLIENT_ID: clientId,
    AZURE_WORD2MD_CLIENT_SECRET: clientSecret,
    AZURE_WORD2MD_REFRESH_TOKEN: refreshToken,
    shareLink, path,
  } = params;
  if (!clientId) {
    return error('AZURE_WORD2MD_CLIENT_ID parameter missing.');
  }
  if (!clientSecret) {
    return error('AZURE_WORD2MD_CLIENT_SECRET parameter missing.');
  }
  if (!refreshToken) {
    return error('AZURE_WORD2MD_REFRESH_TOKEN parameter missing.');
  }
  if (!shareLink) {
    return error('shareLink parameter missing.');
  }
  if (!path) {
    return error('path parameter missing.');
  }

  const drive = new OneDrive({
    clientId,
    clientSecret,
    refreshToken,
  });

  const t0 = Date.now();
  const rootItem = await drive.getDriveItemFromShareLink(shareLink);
  const t1 = Date.now();
  const doc = await drive.getDriveItem(rootItem, path, true);
  const t2 = Date.now();
  const md = await docx2md(doc);
  const t3 = Date.now();
  log.info('resolve=%d, fetch=%d, convert=%d', t1 - t0, t2 - t1, t3 - t2);
  return {
    statusCode: 200,
    body: md,
    headers: {
      'Content-Type': 'text/plain',
      'Content-Length': md.length,
      'Cache-Control': 'no-store, private, must-revalidate',
    },
  };
}

/**
 * Main function
 * @param params Action params
 * @returns {Promise<*>} The response
 */
async function run(params) {
  const { __ow_logger: log } = params;
  try {
    return await fetchViaAPI(params);
  } catch (e) {
    log.error('Error while converting document: %s', e);
    return error('Error while converting document', e.statusCode);
  }
}

/**
 * Main function called by the openwhisk invoker.
 * @param params Action params
 * @returns {Promise<*>} The response
 */
async function main(params) {
  return logger.wrap(run, params);
}

module.exports.main = statusWrap(main, {});
